### PR TITLE
rpc: fix various test bugs

### LINKED ISSUE #108008: rpc: TestRemoteOffsetUnhealthy failed
rpc.TestRemoteOffsetUnhealthy [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11153765?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11153765?buildTab=artifacts#/) on master @ [8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6](https://github.com/cockroachdb/cockroach/commits/8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6):

```
=== RUN   TestRemoteOffsetUnhealthy
I230802 11:41:24.392821 435 2@rpc/peer.go:537  [rnode=2,raddr=127.0.0.1:36277,class=default,rpc] 19  connection is now healthy
I230802 11:41:24.399978 268 2@rpc/peer.go:537  [rnode=3,raddr=127.0.0.1:32905,class=default,rpc] 20  connection is now healthy
I230802 11:41:24.407282 426 2@rpc/peer.go:537  [rnode=4,raddr=127.0.0.1:35415,class=default,rpc] 21  connection is now healthy
I230802 11:41:24.413904 457 2@rpc/peer.go:537  [rnode=1,raddr=127.0.0.1:39329,class=default,rpc] 22  connection is now healthy
I230802 11:41:24.421564 472 2@rpc/peer.go:537  [rnode=3,raddr=127.0.0.1:32905,class=default,rpc] 23  connection is now healthy
I230802 11:41:24.428889 480 2@rpc/peer.go:537  [rnode=4,raddr=127.0.0.1:35415,class=default,rpc] 24  connection is now healthy
I230802 11:41:24.436021 500 2@rpc/peer.go:537  [rnode=1,raddr=127.0.0.1:39329,class=default,rpc] 25  connection is now healthy
I230802 11:41:24.442954 412 2@rpc/peer.go:537  [rnode=2,raddr=127.0.0.1:36277,class=default,rpc] 26  connection is now healthy
I230802 11:41:24.454246 512 2@rpc/peer.go:537  [rnode=4,raddr=127.0.0.1:35415,class=default,rpc] 27  connection is now healthy
I230802 11:41:24.460486 236 2@rpc/peer.go:537  [rnode=1,raddr=127.0.0.1:39329,class=default,rpc] 28  connection is now healthy
I230802 11:41:24.467946 237 2@rpc/peer.go:537  [rnode=2,raddr=127.0.0.1:36277,class=default,rpc] 29  connection is now healthy
I230802 11:41:24.475921 596 2@rpc/peer.go:537  [rnode=3,raddr=127.0.0.1:32905,class=default,rpc] 30  connection is now healthy
    context_test.go:1207: max offset: 100ms - node 0 with acceptable clock offset of 0s did not return an error, as expected
    context_test.go:1207: max offset: 100ms - node 1 with acceptable clock offset of 0s did not return an error, as expected
    context_test.go:1207: max offset: 100ms - node 2 with acceptable clock offset of 0s did not return an error, as expected
    context_test.go:1201: max offset: 100ms - node 3 with excessive clock offset of 100.000001ms returned unexpected error: <nil>
--- FAIL: TestRemoteOffsetUnhealthy (0.11s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoteOffsetUnhealthy.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by knz**
Fixes #108008.

TLDR: Prior to this patch, multiple unit tests in the `rpc` package were misconfiguring their heartbeat interval and timeout. This patch fixes this.

The issue was that there were two separate locations for these tunable parameters:

- `RPCHeartbeat{Interval,Timeout}` fields in `ContextOptions` (previously via `*base.Config`, recently as direct fields in `ContextOptions`)
- also, two private fields `heartbeat{Interval,Timeout}`

The former was copied into the latter during `NewContext()`. Only the latter (the private fields) matter for the actual behavior.

Prior to this commit, the tests would call `NewContext()` and then afterwards would override the fields in `ContextOptions`. These overrides were thus ineffective because the value was already copied by that time.

In addition to this design bug, there were a couple of additional bugs lurking: the overrides defined by the tests were problematic in some cases. If these overrides had been effective, the tests would break.

The problem was discovered after
8d2d2bf47cb3da1c27baa140a5afb5d28a5ed8f6 merged, because in that commit some of the overrides from broken tests were taken on board and became effective. This started causing flakes.

The fix is to remove the ambiguity in configuration and remove the problematic overrides from offending tests.

Release note: None
Epic: CRDB-28893

### CODE REVIEW

**Review by aliher1911 — APPROVED** [review]
Nice untangling exercise!

**Review by pav-kv — APPROVED** [review]

**Comment by knz** [comment]
TFYRs!

bors r=aliher1911,pavelkalinnikov
